### PR TITLE
Fix duplicate assign limit check

### LINKED ISSUE #1593: [Intermediate]: Bot posts spurious `assignment-limit-exceeded` comment after duplicate `/assign` events
### 🧩 Intermediate Friendly

This issue is a good fit for contributors who are already familiar with the Hiero C++ SDK and feel comfortable navigating the codebase.

Intermediate Issues often involve:
- Exploring existing implementations  
- Understanding how different components work together  
- Making thoughtful changes that follow established patterns  

The goal is to support deeper problem-solving while keeping the task clear, focused, and enjoyable to work on.

> [!IMPORTANT]
> ### 🧭 About Intermediate Issues
>
> Intermediate Issues are a great next step for contributors who enjoy digging into the codebase and reasoning about how things work.
>
> These issues often:
> - Involve multiple related files or components  
> - Encourage investigation and understanding of existing behavior  
> - Leave room for thoughtful implementation choices  
> - Stay focused on a clearly defined goal  
>
> Other kinds of contributions — from beginner-friendly tasks to large system-level changes — are just as valuable and use different labels.

### 👾 Description of the Issue

When two `/assign` events for the same issue from the same user fire in quick succession (duplicate webhook delivery, double-posted-then-deleted comment, retry, etc.), the assignment bot:

1. Successfully assigns the contributor on the first run, posts the welcome comment, and swaps `status: ready for dev` → `status: in progress`.
2. Then, ~10–30 seconds later, posts a contradictory `assignment-limit-exceeded` comment on the same issue — telling the contributor they are at the cap and "come back when you're under" — even though the contributor is correctly assigned and within the per-user limit.

The assignment itself is correct. The follow-up comment is wrong and confusing.

**Concrete example:** issue [#1575](https://github.com/hiero-ledger/hiero-sdk-cpp/issues/1575).

- `2026-05-07T04:17:26Z` — `JarvisX-Z` commented `/assign`.
- `2026-05-07T06:41:09Z` — bot assigned `JarvisX-Z`, swapped status labels.
- `2026-05-07T06:41:10Z` — bot posted the welcome comment ([id 4394711468](https://github.com/hiero-ledger/hiero-sdk-cpp/issues/1575#issuecomment-4394711468)).
- `2026-05-07T06:41:31Z` — bot posted the `assignment-limit-exceeded` comment ([id 4394713207](https://github.com/hiero-ledger/hiero-sdk-cpp/issues/1575#issuecomment-4394713207)) saying *"You're currently assigned to **2** open issues. Once you complete or unassign from one of your current issues, come back…"*

GitHub Actions shows two `Bot - On Comment` runs for `JarvisX-Z` on issue #1575, created one second apart and serialized by the per-issue concurrency group:

| Run | Created | Started | Pre-assignment count | Outcome |
|-----|---------|---------|----------------------|---------|
| `25480276791` | `06:40:50Z` | `06:41:03Z` | `1` (only #1429) | Assigned + welcome posted |
| `25480276709` | `06:40:49Z` | `06:41:24Z` | `2` (now includes #1575) | `assignment-limit-exceeded` posted |

The relevant log lines are:

Run `25480276791`:
```
[on-assign] [assign] open assigned issues for JarvisX-Z (excluding blocked): 1
[on-assign] Open assignment count OK: { maxAllowed: 2, currentCount: 1 }
[on-assign] Added assignees: JarvisX-Z
[on-assign] Posted welcome comment
[on-assign] Assignment flow completed successfully
```

Run `25480276709` (~11 seconds later, after concurrency group released):
```
[on-assign] [assign] open assigned issues for JarvisX-Z (excluding blocked): 2
[on-assign] Contributor at or above assignment cap { maxAllowed: 2, currentCount: 2 }
[on-assign] Posted assignment-limit-exceeded comment
```

Relevant files:
```
.github/scripts/commands/assign.js
.github/scripts/commands/assign-comments.js
.github/scripts/tests/test-assign-bot.js
```

### 🔬 Root Cause Analysis

The per-issue concurrency group in [.github/workflows/on-comment.yaml](https://github.com/hiero-ledger/hiero-sdk-cpp/blob/main/.github/workflows/on-comment.yaml) serializes runs on the same issue but does **not** prevent two queued runs from both passing the early gates.

In [`commands/assign.js`](https://github.com/hiero-ledger/hiero-sdk-cpp/blob/main/.github/scripts/commands/assign.js), the gating order in `handleAssign` is:

1. `validateIssueState` — checks `botContext.issue.assignees`, which comes from the **stale webhook payload** captured when the comment was created. Both queued runs see `assignees: []`.
2. `enforceAssignmentLimit` — calls `countIssuesByAssignee`, which hits the **live REST API**. The first run sees `1`; the second run, after the first has already assigned, sees `2` and fails the `>= MAX_OPEN_ASSIGNMENTS` check.
3. `assignAndFinalize` — has a fresh `issues.get()` pre-flight that defends against this exact race for the *same-issue, different-user* case (see comments at lines 442–475), but the second run never reaches it because `enforceAssignmentLimit` fails first.

The result: the `assignAndFinalize` race defense is bypassed by an earlier gate that reads live state, and the user sees a contradictory comment.

### ✅ Expected Behavior

When two `/assign` events for the same `(user, issue)` pair fire in quick succession:

- Exactly one welcome comment is posted.
- The contributor is assigned exactly once.
- No `assignment-limit-exceeded` comment is posted, because the contributor is correctly within the per-user limit (the count of `2` includes the issue they were just assigned to — they are not over the cap).

The second run should detect that the requester is already the assignee of this issue and exit silently (or at most post an idempotent already-assigned acknowledgement, matching `buildAlreadyAssignedComment`).

### ❌ Actual Behavior

The bot posts both a welcome comment and an `assignment-limit-exceeded` comment ~10–30 seconds apart on the same issue. Contributors see contradictory messages and are unsure whether they actually have the issue or need to drop one of their other assignments.

### 🌐 Environment

- Repo: `hiero-ledger/hiero-sdk-cpp`
- Workflow: `.github/workflows/on-comment.yaml`
- Bot scripts: `.github/scripts/commands/assign.js`, `.github/scripts/helpers/api.js`
- Node runtime: as configured in `actions/github-script@v9`
- First observed: 2026-05-07 on issue #1575 (`JarvisX-Z`)

### 💡 Proposed Approach

Two reasonable approaches; the contributor should pick one and explain the trade-off in the PR.

**Option A — fresh-state pre-flight in `enforceAssignmentLimit`:** before posting `buildAssignmentLimitExceededComment`, fetch fresh issue state via `issues.get()` and check whether `requesterUsername` is already in `issue.assignees`. If so, treat it as a duplicate trigger and return without posting (or post `buildAlreadyAssignedComment`). This mirrors the existing same-issue race defense in `assignAndFinalize` (lines 442–475).

**Option B — move the duplicate-trigger check into `validateIssueState`:** replace the stale-payload `botContext.issue.assignees` check with a fresh `issues.get()`. This consolidates same-issue race protection into a single early gate and removes the need for a second pre-flight in `assignAndFinalize`. Slightly larger refactor but eliminates the inconsistency between the two gates' freshness.

Either way, the limit-exceeded path must not fire when the requester is already the assignee of the current issue.

### 👩‍💻 Implementation Steps

- [ ] Reproduce the bug in `test-assign-bot.js` by simulating two queued `/assign` runs for the same `(user, issue)` pair where the live REST count returns `1` then `2`. Confirm the current code posts both a welcome comment and an `assignment-limit-exceeded` comment.
- [ ] Decide between Option A and Option B (or propose a third). Document the choice in the PR description.
- [ ] Implement the fix in [.github/scripts/commands/assign.js](https://github.com/hiero-ledger/hiero-sdk-cpp/blob/main/.github/scripts/commands/assign.js).
- [ ] Verify that the existing same-issue, different-user race defense in `assignAndFinalize` still works (the `freshIssue.assignees?.length > 0` branch must still post `buildAlreadyAssignedComment` when a *different* user was assigned mid-queue).
- [ ] Add a regression test in [.github/scripts/tests/test-assign-bot.js](https://github.com/hiero-ledger/hiero-sdk-cpp/blob/main/.github/scripts/tests/test-assign-bot.js) covering: (a) duplicate trigger, same user — only one welcome comment, no limit-exceeded; (b) same-issue race, different users — `buildAlreadyAssignedComment` still fires for the loser.
- [ ] Run `npm test` from `.github/scripts/` and confirm all assign-bot tests pass.
- [ ] Verify legitimate over-limit behavior is preserved: a user with `2` other open issues running `/assign` on a third should still receive `buildAssignmentLimitExceededComment`.

### ✔️ Acceptance Criteria

- [ ] Two queued `/assign` events for the same `(user, issue)` pair result in one welcome comment and zero `assignment-limit-exceeded` comments.
- [ ] A user genuinely at the assignment cap (e.g. already assigned to two other open, non-blocked issues) still receives `buildAssignmentLimitExceededComment` when they run `/assign` on a new issue.
- [ ] Same-issue race between two different users still posts `buildAlreadyAssignedComment` for the second user.
- [ ] A regression test covers the duplicate-trigger same-user case.
- [ ] All existing tests in `test-assign-bot.js` continue to pass.
- [ ] No unrelated behavior or API changes are introduced.

---

### 📋 Step-by-Step Contribution Guide

To help keep contributions consistent and easy to review, we recommend following these steps:

- [ ] Comment `/assign` to request the issue
- [ ] Wait for assignment
- [ ] Fork the repository and create a branch
- [ ] Set up the project using the instructions in `README.md`
- [ ] Make the requested changes
- [ ] Sign each commit using `-s -S`
- [ ] Push your branch and open a pull request

Read [Workflow Guide](https://github.com/hiero-ledger/hiero-sdk-cpp/blob/main/docs/training/workflow.md) for step-by-step workflow guidance.
Read [README.md](https://github.com/hiero-ledger/hiero-sdk-cpp/blob/main/README.md) for setup instructions.

❗ Pull requests **cannot be merged** without `S` and `s` signed commits.
See the [Signing Guide](https://github.com/hiero-ledger/hiero-sdk-cpp/blob/main/docs/training/signing.md).

### 🤔 Additional Information

If you have questions while working on this issue, feel free to ask!

You can reach the community and maintainers here: [Hiero-SDK-C++ Discord](https://discord.com/channels/905194001349627914/1337424839761465364)

Whether you need help finding the right file, understanding existing code, or confirming your approach — we're happy to help.

### PR DESCRIPTION
**Pull request by hk2166**
**Description**:
Fix duplicate `/assign` handling so a queued duplicate event does not post a spurious assignment-limit-exceeded comment after the requester has already been assigned.

* Add a fresh issue-state check before posting the assignment-limit-exceeded comment
* Return the existing already-assigned response when the requester is already assigned to the current issue
* Add a regression test for duplicate same-user `/assign` events at the assignment cap

**Related issue(s)**:

Fixes #1593

### CODE REVIEW

**Comment by hk2166** [comment]
Hi @rwalworth @SimiHunjan  @agadzhalov 

I fixed this by adding a fresh issue-state check before the bot posts the assignment-limit-exceeded comment.

If the live issue data shows that the requester is already assigned to the same issue, the bot treats the run as a duplicate `/assign` event and returns the existing already-assigned response instead of posting the limit warning.

This keeps the normal over-limit behavior unchanged for users who are actually trying to assign a new issue while already at the cap.

I also added a regression test for the duplicate `/assign` case and verified the assign-bot test suite locally.

Please review it when you get a chance.

File: .github/scripts/commands/assign.js [hunk 1]
```diff
@@ -329,6 +329,59 @@ async function enforceAssignmentLimit(botContext, requesterUsername) {
...
+        botContext,
+        buildAlreadyAssignedComment(
+          requesterUsername,
+          freshIssue,
+          botContext.owner,
+          botContext.repo,
+        ),
+      );
+      return false;
+    }
```
Comment (rwalworth) [inline .github/scripts/commands/assign.js]: **issue (blocking):** This second branch handles a case the linked issue didn't ask you to change: when the requester is at the cap AND a different user got assigned mid-queue. Previously that path posted `buildAssignmentLimitExceededComment`; now it posts `buildAlreadyAssignedComment` ("This issue is already assigned to @other-user… find another issue to work on").

A few concerns with that:

1. The acceptance criteria included *"No unrelated behavior or API changes are introduced"* - and this changes behavior for the at-cap + snatched-by-other-user case, which the bug didn't touch. The existing same-issue race defense in `assignAndFinalize` (lines 478–490) already handles this case when the user is *not* at the cap.
2. The advice in the new comment ("find another issue to work on") is misleading when the user is actually at their assignment cap - they can't take another issue either way.
3. The new branch isn't covered by a test scenario (the new test only exercises the same-user duplicate path).
4. The leading comment `// Suppress limit failures caused by a duplicate /assign for this issue.` only describes the `requesterAlreadyAssigned` branch, not this one.

Please remove this branch so the change stays tightly scoped to the actual bug - the multi-user race already lives in `assignAndFinalize` and doesn't need a second defense here. Once that's out, I'll re-review and merge.

**Review by rwalworth — CHANGES_REQUESTED** [review]
Nice work @hk2166, the core fix is correct and the regression test covers the bug scenario well. Verified all 42 assign-bot tests pass locally and ESLint is clean. One change needed before I can merge - I left a comment below about a second branch in `enforceAssignmentLimit` that changes behavior outside the scope of the bug. Let me know if you have any questions!

File: .github/scripts/commands/assign.js (hunk 1 above)
Comment (hk2166) [reply .github/scripts/commands/assign.js]: Thanks for the review. I removed the extra freshIssue.assignees?.length > 0 branch from enforceAssignmentLimit, so the change is now scoped to the same-user duplicate /assign case only. I also re-ran the assign bot tests and ESLint locally: 42/42 tests pass and ESLint is clean. Could you please re-review when you get a chance?

**Review by rwalworth — APPROVED** [review]
Thanks for the quick turnaround @hk2166 - the second branch is removed, the leading comment now accurately describes the remaining check, and all 42 assign-bot tests still pass locally. The fix is tightly scoped to the actual bug now. Merging!
